src/backend/metal/src/command.rs

```diff
@@ -14,416 +14,112 @@
-            }
+        if !self.encoder.is_render_encoding() {
+            self.encoder.begin_render_encoding();
         }
 
-        use map::map_index_type;
-
-        match draw {
-            Draw::Normal(count, start) => {
-                encoder.draw_primitives(MTLPrimitiveType::Triangle, start, count)
-            }
-            Draw::Instanced(count, ninst, start, _offset) => {
-                encoder.draw_primitives_instanced(MTLPrimitiveType::Triangle, start, count, ninst);
-            }
-            Draw::Indexed(count, start, _base) => {
-                encoder.draw_indexed_primitives(MTLPrimitiveType::Triangle,
-                                                count,
-                                                map_index_type(self.index_buf.unwrap().1),
-                                                ((self.index_buf.unwrap().0).0).0,
-                                                start);
-            }
-            Draw::IndexedInstanced(_count, _ninst, _start, _base, _offset) => unimplemented!(),
+        if self.should_restore {
+            self.encoder.restore_render_state();
+            self.should_restore = false;
         }
-
-        // self.cache.clear();
-        // self.buf.clear();
     }
 }
 
 impl command::Buffer<Resources> for CommandBuffer {
     fn reset(&mut self) {
-        self.cache.clear();
-        self.buf.clear();
+        self.encoder.reset();
     }
 
     fn bind_pipeline_state(&mut self, pso: Pipeline) {
-        self.buf.push(Command::BindPipeline(pso));
+        self.encoder.set_render_pipeline_state(pso.pipeline);
+        self.encoder.set_front_facing_winding(pso.winding);
+        self.encoder.set_cull_mode(pso.cull);
+        self.encoder.set_triangle_fill_mode(pso.fill);
+
+        // TODO(fkaa): do we need max value?
+        self.encoder.set_depth_bias(pso.depth_bias as f32 ,
+                                    pso.slope_scaled_depth_bias as f32,
+                                    0f32);
+
+        if let Some(depth_state) = pso.depth_stencil {
+            self.encoder.set_depth_stencil_state(depth_state);
+        }
     }
 
     fn bind_vertex_buffers(&mut self, vbs: pso::VertexBufferSet<Resources>) {
-        let mut buffers = [native::Buffer(MTLBuffer::nil()); MAX_VERTEX_ATTRIBUTES];
-        let mut offsets = [0; MAX_VERTEX_ATTRIBUTES];
-        let mut indices = [0; MAX_VERTEX_ATTRIBUTES];
-
-        for i in 0..1 {
+        for i in 0..MAX_VERTEX_ATTRIBUTES {
             if let Some((buffer, offset)) = vbs.0[i] {
-                buffers[i] = buffer.0;
-                offsets[i] = offset as u64;
-                indices[i] = i as u64;
+                // TODO(fkaa): is it okay if we cascade downwards with vertex
+                //             buffers? don't want to collide with user defined
+                //             values on accident!
+                self.encoder.set_vertex_buffer(30 - i as u64, offset as u64, unsafe { *(buffer.0).0 });
```

Comment:
Forgot to use `MTL_MAX_BUFFER_BINDINGS` (31). Otherwise is it OK that vertex buffers (not mentioned in shaders, only PSO macro) are automatically assigned from top to bottom? would be unpleasant if you bind your own buffers at say 0 & 1 and it collides with another buffer. Should still warn/error at runtime though.